core/src/main/java/org/apache/iceberg/SchemaUpdate.java

```diff
@@ -340,26 +407,30 @@ public Type field(Types.NestedField field, Type fieldResult) {
     }
 
     @Override
-    public Type list(Types.ListType list, Type result) {
+    public Type list(Types.ListType list, Type elementResult) {
       // use field to apply updates
-      Type elementResult = field(list.fields().get(0), result);
-      if (elementResult == null) {
+      Types.NestedField elementField = list.fields().get(0);
+      Type elementType = field(elementField, elementResult);
+      if (elementType == null) {
         throw new IllegalArgumentException("Cannot delete element type from list: " + list);
       }
 
-      if (list.elementType() == elementResult) {
+      Types.NestedField elementUpdate = updates.get(elementField.fieldId());
+      boolean isElementOptional = elementUpdate != null ? elementUpdate.isOptional() : list.isElementOptional();
+
+      if (isElementOptional == elementField.isOptional() && list.elementType() == elementType) {
         return list;
       }
 
-      if (list.isElementOptional()) {
-        return Types.ListType.ofOptional(list.elementId(), elementResult);
+      if (isElementOptional) {
+        return Types.ListType.ofOptional(list.elementId(), elementType);
```

Comment:
Thoughts on allowing doc on list field elements, maybe as part of a separate rb?